runtime/lua/vim/lsp/util.lua

```diff
@@ -1290,18 +1290,159 @@ function M._trim(contents, opts)
   return contents
 end
 
---- Generates a table mapping markdown code block lang to vim syntax,
---- based on g:markdown_fenced_languages
----@return table table of lang -> syntax mappings
-local function get_markdown_fences()
-  local fences = {}
-  for _, fence in pairs(vim.g.markdown_fenced_languages or {}) do
-    local lang, syntax = fence:match('^(.*)=(.*)$')
-    if lang then
-      fences[lang] = syntax
-    end
-  end
-  return fences
+---@alias MarkdownBlock {line:string}
+---@alias MarkdownCodeBlock {code:string[], lang:string}
+---@alias Markdown (MarkdownBlock|MarkdownCodeBlock)[]
+
+---@param text string
+---@param opts? {ft?: string}
+---@return Markdown
+local function parse_markdown(text, opts)
+  opts = opts or {}
+  text = text:gsub('</?pre>', '```'):gsub('\r', '')
+
+  -- Replace common HTML entities.
+  local entities =
+    { nbsp = '', lt = '<', gt = '>', amp = '&', quot = '"', apos = "'", ensp = ' ', emsp = ' ' }
+  for entity, char in pairs(entities) do
+    text = text:gsub('&' .. entity .. ';', char)
+  end
+
+  local ret = {} ---@type Markdown
+  local lines = vim.split(text, '\n')
+  local l = 1
+
+  local function is_empty(line)
+    return line and line:find('^%s*$')
+  end
+
+  local function eat_nl()
+    while is_empty(lines[l + 1]) do
+      l = l + 1
+    end
+  end
```

Comment:
Suggestion: instead of using this function for side effects, pass in current line and return modified line:

```suggestion
  local function eat_nl(l)
    while is_empty(lines[l + 1]) do
      l = l + 1
    end
    return l
  end
```

Then at the call site you'd use:

```lua
l = eat_nl(l)
```

Reply:
Just curious: Why don't you like the closure modifying the outer variable?

Reply:
Looking at a naked function call like:

```lua
eat_nl()
```

it is not obvious what side effects it has. It certainly is not clear that it modifies the local `l` variable without looking at the implementation. Someone reading this code should be able to follow the modifications to the variables as they go. This makes it easier to reason about and debug.

Generally speaking, "pure" functions (whose outputs only depend on their inputs, with no side effects) should be preferred, where possible and reasonable to do so.